### DISCUSSION TITLE
Use of `:=` operator in optional function arguments.

### DISCUSSION THREAD

**Discussion by jgrg** [duckdb/duckdb#13791] [Q&A] [answered]
After coming across the `:=` operator in the `unnest()` function:
```sql
SELECT unnest([{'a': 42, 'b': 84}, {'a': 100, 'b': NULL}], recursive := true);
```

I tried to find out about the `:=` operator and what its significance is, but it isn't defined in the documentation. (And you can't search for it on the website.) It is otherwise used for building structs, unions, and setting default values of arguments in macro definitions.

As far as I can see, other functions use `=` for optional arguments, _e.g._
```sql
FROM read_csv(
    'faulty.csv',
    columns = {'name': 'VARCHAR', 'age': 'INTEGER'},
    store_rejects = true
);
```

Are there any other functions which use `:=` for optional arguments?

Why is `:=` used in `unnest()`?

**Comment by Maxxen** [comment, marked as answer]
If I recall correctly the reason is that `:=` is used in scalar functions to disambiguate from the `=` equality comparison operator. Table functions (like read_csv) can't take column expressions as arguments anyway so for them there is no ambiguity. 

To illustrate:

```sql
CREATE FUNCTION my_func(x := 1337) AS x + 1;

SELECT my_func();
┌───────────┐
│ my_func() │
│   int32   │
├───────────┤
│      1338 │
└───────────┘

SELECT my_func(x := 10);
┌──────────────────┐
│ my_func(x := 10) │
│      int32       │
├──────────────────┤
│               11 │
└──────────────────┘

-- Now lets create a table
CREATE TABLE t1 AS SELECT 10 AS x;

-- This won't work, but image it did
SELECT my_func(x = 10) FROM t1 
-- ^does this evaluate to `my_func(true)` => 2 or `my_func(x)` => 11?
```

**Reply by Tishj** [reply]
This behaves similarly to the `=>` variant in Postgres, which we will also support starting at 1.1.0
<https://github.com/duckdb/duckdb/pull/12913>

**Reply by jgrg** [reply]
OK, but `unnest()` is a table function:
```sql
FROM duckdb_functions() WHERE function_name = 'unnest';
```
```
   database_name = system
    database_oid = 0
     schema_name = main
   function_name = unnest
   function_type = table
     description =
         comment =
            tags = {}
     return_type =
      parameters = [col0]
 parameter_types = [TABLE]
         varargs =
macro_definition =
has_side_effects =
        internal = true
    function_oid = 86
         example =
       stability =
```

**Reply by Maxxen** [reply]
Not only. Its a table function but there is also a special magic scalar version, which is what you use when you do "SELECT unnest()". The table version is "SELECT * FROM unnest()"

**Comment by jgrg** [comment]
OK, thanks.
I still haven't been able to find any other functions which use `:=` for optional arguments.

**Comment by Maxxen** [comment]
In DuckDB 2.0 all function will be able to pass arguments by name https://github.com/duckdb/duckdb/pull/22941
